### PR TITLE
Vs refactor parser cleanup

### PR DESCRIPTION
**Pull request by not-napoleon**
This PR removes lingering hard coded type logic in `ValuesSourceParserHelper`, and tries to minimize the exposure of the `ValueType.isA` concept, which has caused some weird casting situations before.

Relates to #42949

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridAggregationBuilder.java [hunk 1]
```diff
@@ -65,7 +64,7 @@
 
     public static ObjectParser<GeoGridAggregationBuilder, Void> createParser(String name, PrecisionParser precisionParser) {
         ObjectParser<GeoGridAggregationBuilder, Void> parser = new ObjectParser<>(name);
-        ValuesSourceParserHelper.declareGeoFields(parser, false, false);
+        ValuesSourceAggregationBuilder.declareFields(parser, false, false, false);
```
Comment (nik9000) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridAggregationBuilder.java]: Probably a thing for later more than now: three boolean parameters can be kind of hard to keep track of. I think IntelliJ has something where it puts the names of parameters but that isn't universal. Anyway, *something* else might be a little more descriptive. I'm not sure what though. It certainly can wait for later too.

Comment (not-napoleon) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/geogrid/GeoGridAggregationBuilder.java]: I went back and forth on that.  The other obvious alternative is to have a `declareTimeZoneAwareField` and `declareTimeZoneUnawareField` methods, which would replace the third boolean.  But I kind of hate moving parameters into method names, so I went with this one.  I agree it's hard to keep track of though.  I suppose we could make it a small builder, but that seems like too much.  IDK.  Open to suggestions.

**Comment by not-napoleon** [comment]
@elasticmachine run elasticsearch-ci/2

File: server/src/main/java/org/elasticsearch/search/aggregations/support/CoreValuesSourceType.java [hunk 2]
```diff
@@ -291,24 +290,6 @@ public DocValueFormat getFormatter(String format, ZoneId tz) {
     }
     ;
 
-    enum EquivalenceType {
```
Comment (polyfractal) [inline server/src/main/java/org/elasticsearch/search/aggregations/support/CoreValuesSourceType.java]: :tada:

**Review by polyfractal — APPROVED** [review]
LGTM!

**Pull request merged**